File: doc/back/pow.m

     NAME
          exp, log, log10, pow, sqrt - exponential, logarithm, power,
          square root functions

     SYNTAX
          #include <math.h>

          double exp (x)
          double x;

          double log (x)
          double x;

          double log10 (x)
          double x;

          double pow (x, y)
          double x, y;

          double sqrt (x)
          double x;

     DESCRIPTION
          Exp returns e**x.

          Log returns the natural logarithm of x.  The value of x must
          be positive.

          Log10 returns the logarithm base ten of x.  The value of x
          must be positive.

          Pow returns x**y.  If x is zero, y must be positive.  If x is
          negative, y must be an integer.

          Sqrt returns the non-negative square root of x.  The value
          of x may not be negative.

     DIAGNOSTICS
          Exp returns HUGE when the correct value would overflow, or 0
          when the correct value would underflow, and sets errno to
          ERANGE.

          Log and log10 return -HUGE and set errno to EDOM when x is
          non-positive.  A message indicating DOMAIN error (or SING
          error when x is 0) is printed on the standard error output.

          Pow returns 0 and sets errno to EDOM when x is 0 and y is
          non-positive, or when x is negative and y is not an integer.
          In these cases a message indicating DOMAIN error is printed
          on the standard error output.  When the correct value for
          pow would overflow or underflow, pow returns +HUGE or 0
          respectively, and sets errno to ERANGE.

          Sqrt returns 0 and sets errno to EDOM when x is negative.  A
          message indicating DOMAIN error is printed on the standard
          error output.

          These error-handling procedures may be changed with the
          function matherr().

     SEE ALSO
          hypot.m, matherr.m, sinh.m

          (printed 07/29/92 - J B Systems)

File: nblib/pow.c
/*~!pow.c*/
/* Name:  pow.c Part No.: _______-____r
 *
 * Copyright 1992 - J B Systems, Morrison, CO
 *
 * The recipient of this product specifically agrees not to distribute,
 * disclose, or disseminate in any way, to any one, nor use for its own
 * benefit, or the benefit of others, any information contained  herein
 * without the expressed written consent of J B Systems.
 *
 *                     RESTRICTED RIGHTS LEGEND
 *
 * Use, duplication, or disclosure by the Government is  subject  to
 * restriction  as  set forth in paragraph (b) (3) (B) of the Rights
 * in Technical Data and Computer Software  Clause  in  DAR  7-104.9
 * (a).
 */

#ident	"@(#)nbclib:pow.c	1.1"

/*
 *	pow(x, y) returns x ** y.
 *	Returns EDOM error and value 0 for 0 to a non-positive power
 *	or negative to a non-integral power;
 *	ERANGE error and value HUGE or -HUGE when the correct value
 *	would overflow, or 0 when the correct value would underflow.
 *	uses log and exp
 *	This version accepts negative x and integral y,
 *	apparently in violation of the ANSI FORTRAN standard for x ** y.
 *	There is a much more accurate but much more elaborate algorithm
 *	in Cody and Waite, which should be substituted for this.
 */

#include <math.h>
#include <values.h>
#include <errno.h>

double
pow(x, y)
register double x, y;
{
	register int neg;
	struct exception exc;

	if (y == 1) /* easy special case */
		return (x);
	exc.name = "pow";
	exc.arg1 = x;
	exc.arg2 = y;
	exc.retval = 0.0;
	if (!x) {
		if (y > 0)
			return (x); /* (0 ** pos) == 0 */
		goto domain;
	}
	neg = 0;
	if (x < 0) { /* test using integer arithmetic if possible */
		if (y >= -MAXLONG && y <= MAXLONG) {
			register long ly = (long)y;

			if ((double)ly != y)
				goto domain; /* y not integral */
			neg = ly % 2;
		} else {
			double fr, dum, modf();

			if (fr = modf(0.5 * y, &dum)) {
				if (fr != 0.5)
					goto domain; /* y not integral */
				neg++; /* y is odd */
			}
		}
		x = -x;
	}
	if (x != 1) { /* x isn't the final result */
		/* the following code protects against multiplying x and y
		 * until there is no chance of multiplicative overflow */
		if ((x = log(x)) < 0) { /* preserve sign of product */
			x = -x;
			y = -y;
		}
		if (y > LN_MAXDOUBLE/x) {
			exc.type = OVERFLOW;
			exc.retval = neg ? -HUGE : HUGE;
			if (!matherr(&exc))
				errno = ERANGE;
			return (exc.retval);
		}
		if (y < LN_MINDOUBLE/x) {
			exc.type = UNDERFLOW;
			if (!matherr(&exc))
				errno = ERANGE;
			return (exc.retval);
		}
		x = exp(x * y); /* finally; no mishap can occur */
	}
	return (neg ? -x : x);

domain:
	exc.type = DOMAIN;
	if (!matherr(&exc)) {
		(void) write(2, "pow: DOMAIN error\n", 18);
		errno = EDOM;
	}
	return (exc.retval);
}
